typing/printtyp.ml

```diff
@@ -1129,28 +1129,40 @@ let rec tree_of_typexp mode ty =
                | _ -> false)
             fields in
         let all_present = List.length present = List.length fields in
-        begin match name with
-        | Some(p, tyl) when nameable_row row ->
-            let (p', s) = best_type_path p in
-            let id = tree_of_best_type_path p p' in
-            let args = tree_of_typlist mode (apply_subst s tyl) in
-            let out_variant =
-              if is_nth s then List.hd args else Otyp_constr (id, args) in
-            if closed && all_present then
-              out_variant
-            else
-              let non_gen = is_non_gen mode (Transient_expr.type_expr px) in
+        let result, non_gen =
+          match name with
+          | Some(p, tyl) when nameable_row row ->
+              let (p', s) = best_type_path p in
+              let id = tree_of_best_type_path p p' in
+              let args = tree_of_typlist mode (apply_subst s tyl) in
+              let out_variant =
+                if is_nth s then List.hd args else Otyp_constr (id, args) in
+              if closed && all_present then
+                out_variant, false
+              else
+                let non_gen = is_non_gen mode (Transient_expr.type_expr px) in
+                let tags =
+                  if all_present then None else Some (List.map fst present) in
+                Otyp_variant (non_gen, Ovar_typ out_variant, closed, tags)
+              , non_gen
+          | _ ->
+              let non_gen =
+                not (closed && all_present) &&
+                is_non_gen mode (Transient_expr.type_expr px) in
+              let fields = List.map (tree_of_row_field mode) fields in
               let tags =
                 if all_present then None else Some (List.map fst present) in
-              Otyp_variant (non_gen, Ovar_typ out_variant, closed, tags)
-        | _ ->
-            let non_gen =
-              not (closed && all_present) &&
-              is_non_gen mode (Transient_expr.type_expr px) in
-            let fields = List.map (tree_of_row_field mode) fields in
-            let tags =
-              if all_present then None else Some (List.map fst present) in
-            Otyp_variant (non_gen, Ovar_fields fields, closed, tags)
+              Otyp_variant (non_gen, Ovar_fields fields, closed, tags)
+            , non_gen
+        in
+        begin match non_gen, (Transient_expr.repr more).desc, mode with
+        | true, Tvar _, Type_scheme ->
+            Otyp_alias
+              ( result
+              , "_" ^ Names.(name_of_type
+                               (new_weak_name more)
+                               (Transient_expr.repr more)))
```

Comment:
Creating an alias by hand will interact poorly with the aliasing mechanism in `Printtyp`. For instance,
```ocaml
let rec x = `A (`B (fun y -> x = y))
module M = struct
  let y = Fun.id x
end
```
currently yield the following error
```
Error: The type of this module,
       sig
         val y : _[> `A of _[> `B of 'a -> bool ] as '_weak1 ] as '_a as 'a
       end, contains the non-generalizable type variable(s) '_a, '_weak1
       (See manual section 6.1.2)
File "test.ml", line 3, characters 6-7:
3 |   let y = Fun.id x
          ^
  This item has the non-generalizable type 'a.
```

Reply:
I'm trying to use user-supplied names instead of `_weakN` if available, since it's probably more clear this way. 
Do you have any suggestions on how to do it properly?

Reply:
See #12107 for an implementation of the `_ as _weak` idea that reuses the currently existing aliasing mechanism to avoid adding a second layer of aliases.